Add schema type metadata (#231)
oauth-wg/oauth-sd-jwt-vc: draft-ietf-oauth-sd-jwt-vc.md

This specification uses the terms "Holder", "Issuer", "Verifier", "Key Binding", and "Key Binding JWT" defined by
[@!I-D.ietf-oauth-selective-disclosure-jwt].

Consumer:

> : Applications using this specification are called Consumer. This typically includes Issuers, Verifiers, and Wallets.

Reviewer: ```suggestion
: Applications using the type metadata specified in (#type-metadata) are called Consumer. This typically includes Issuers, Verifiers, and Wallets.
```

Author: Thanks, I wanted to change this as well :)

Reviewer 2: Thank you, yes, this needed the additional context.

Furthermore, the recipient of the SD-JWT VC MUST validate the public verification key
for the Issuer-signed JWT as defined in (#issuer-signed-jwt-verification-key-validation).

> If Schema Type Metadata is provided, a recipient MUST validate the JSON Schema as defined in (#schema-type-metadata).

Reviewer: "Schema" should only be uppercased when used in "JSON Schema", otherwise I'd use lowercase (unless we define the term).

# Type Metadata {#type-metadata}

> A SD-JWT VC type, i.e., the `vct` value, is associated with Type Metadata defining, for example, information about the type or a schema defining which claims MAY or MUST appear in the SD-JWT VC.

Reviewer: ```suggestion
An SD-JWT VC type, i.e., a `vct` value, is associated with Type Metadata defining, for example, information about the type or a schema defining which claims MAY or MUST appear in the SD-JWT VC.
```

This section defines Type Metadata that can be associated with a type of a SD-JWT VC, as well as a method for retrieving the Type Metadata and processing rules. This Type Metadata is intended to be used, among other things, for the following purposes:

 * Developers of Issuers and Verifiers can use the Type Metadata to understand the
   semantics of the type and the associated rules. While in some cases,
   Issuers are the parties that define types, this is
   not always the case. For example, a type can be defined by a
   standardization body or a community.
 * Verifiers can use the Type Metadata to determine whether a credential is valid
   according to the rules of the type. For example, a Verifier can check
   whether a credential contains all required claims and whether the claims
   are selectively disclosable.

Type Metadata can be retrieved as described in (#retrieving-type-metadata).

## Type Metadata Example {#type-metadata-example}

All examples in this section are non-normative.

```json
{

>   "vct": "https://betelgeuse.example.com/education_credential",

Reviewer 3: Is the `vct` field required in a metadata JSON object? This would break compatibility with content addressable storage systems. For example, if I want to use `ipfs://<hash>` as a `vct` uri. It would be easy to use the ipfs-uri in the JWT, but it's impossible to put it inside of the metadata JSON object, since it would require knowing the hash of that object before calculating it.

Seems like `vct` needs to be optional here.

Author: Yes, I agree that this would not work with content addressable storage systems the way it is defined now. It is a chicken and egg problem. IMO, we should support content addressable hash URIs for `vct` and make `vct` optional in the type metadata.

@danielfett @bc-pi is there is a security-related reason to make `vct` in type metadata mandatory?

Reviewer: As it is designed at the moment, vct is at least required for usage in glue documents. We need to figure out how to address this.

Reviewer 2: I'm not sure that storage and retrieval of Type Metadata from content addressable storage systems is or should a goal.

Author: IMO, there is a community that uses this and SD-JWT VC and if possible we should help them to support this.

Reviewer 3: > I'm not sure that storage and retrieval of Type Metadata from content addressable storage systems is or should a goal.

This makes me sad. If sd-jwt vc actively is trying to disallow usage by CAS systems we won't be able to adopt it. 
There are a few reasons why I think using CAS (over something like http) is nice:
1. Less trust assumptions, if the url is the hash, I know what I get is correct no matter who gives it to me
2. Less technical assumptions, hashes are generally technically simpler than https signatures
3. Content can't be arbitrarily changed, with https the DNS controller ultimately decides what content is served

Author: The goal should be to not prohibit it, we don't need to add normative language around that.

Reviewer: As this question is not directly related to the PR, please continue the discussion in this issue: https://github.com/oauth-wg/oauth-sd-jwt-vc/issues/211

Reviewer 3: @danielfett I would say it's actually directly related ro the PR since the issue only exist in the context of the `vct` field inside of the type metadata object.

# Type Metadata {#type-metadata}

> A SD-JWT VC type, i.e., the `vct` value, is associated with Type Metadata defining, for example, information about the type or a Schema defining (see (#schema-definition)) which claims MAY or MUST appear in the SD-JWT VC.

Reviewer: "Schema" as above

This section defines Type Metadata that can be associated with a type of a SD-JWT VC, as well as a method for retrieving the Type Metadata and processing rules. This Type Metadata is intended to be used, among other things, for the following purposes:

 * Developers of Issuers and Verifiers can use the Typ Metadata to understand the
 * Developers of Issuers and Verifiers can use the Type Metadata to understand the
   semantics of the type and the associated rules. While in some cases,
   Issuers are the parties that define types (credential formats), this is
   Issuers are the parties that define types, this is
   not always the case. For example, a type can be defined by a
   standardization body or a community.
 * Verifiers can use the metadata to determine whether a credential is valid
 * Verifiers can use the Type Metadata to determine whether a credential is valid
   according to the rules of the type. For example, a Verifier can check
   whether a credential contains all required claims and whether the claims
   are selectively disclosable.
 * Wallets can use the metadata to display the credential in a way that is

>    consistent with the Issuer's intent.

Reviewer 2: is this not true? or moved?

Author: Not yet :) because the display feature is still pending

Reviewer 2: sigh... I really dislike how we've elected to do this. Is that bullet thereby included in the PR for that feature?

> Type Metadata can be retrieved as described in (#retrieving-type-metadata).

Reviewer 2: This change effectively removes the definition of "Consumer" but the term is used a few times later on.

Reviewer 2: nevermind, I see it moved

* `extends`
  * OPTIONAL. A URI of another type that this type extends, as described in
  (#extending-type-metadata).
* `schema`
  * OPTIONAL. An embedded JSON Schema document describing the structure of
  the Verifiable Credential as described in (#schema-type-metadata). `schema` MUST NOT be used
  if `schema_uri` is present.
* `schema_uri`
  * OPTIONAL. A URL pointing to a JSON Schema document describing the structure
  of the Verifiable Credential as described in (#schema-type-metadata). `schema_uri` MUST NOT

>   be used if `schema` is present.

Reviewer 2: are both really needed?

Reviewer: I think so, yes. Some schemas already exist, they can be referenced here. In other cases, embedding is useful.

   (#document-integrity). The Consumer MUST NOT use the Type Metadata if no hash for integrity protection was provided.

## Extending Type Metadata {#extending-type-metadata}

A SD-JWT VC type can extend another type. The extended type is identified by the URI in
the `extends` property. Consumers MUST retrieve and process
Type Metadata for the extended type before processing the Type Metadata for the extending
type.

The extended type MAY itself extend another type. This can be used to create a
chain or hierarchy of types. The security considerations described in
(#circular-extends) apply in order to avoid problems with circular dependencies.

The `schema` or `schema_uri` property of the extending type MUST be evaluated for validation purposes if it exists.
Afterwards, or if there is no such property, the `schema` or `schema_uri` property of the extended type MUST evaluated.
This means that all schemas in the chain of extended types MUST be evaluated for a particular Verifiable Credential.

## Schema Type Metadata {#schema-type-metadata}

> If a `schema` or `schema_uri` property is present, a Consumer that intends to validate the Verifiable Credential

Reviewer 2: why both?

   Type Metadata document matches that of the `vct` in the Type Metadata document, and it MUST verify the integrity of the Type Metadata document as defined in
   (#document-integrity). The Consumer MUST NOT use the Type Metadata if no hash for integrity protection was provided.

## Extending Type Metadata {#extending-type-metadata}

> A SD-JWT VC type can extend another type. The extended type is identified by the URI in

Reviewer: ```suggestion
An SD-JWT VC type can extend another type. The extended type is identified by the URI in
```

the `extends` property. Consumers MUST retrieve and process
Type Metadata for the extended type before processing the Type Metadata for the extending
type.

The extended type MAY itself extend another type. This can be used to create a
chain or hierarchy of types. The security considerations described in
(#circular-extends) apply in order to avoid problems with circular dependencies.

## Schema Type Metadata {#schema-type-metadata}

### Schema Definition {#schema-definition}

Schemas for Verifiable Credentials are contained in the `schema` or retrieved via the `schema_uri` Type Metadata parameters (as defined in (#type-metadata-format)).

> A Schema MUST be represented by a JSON Schema document according to draft version 2020-12 [@JSON.SCHEMA.2020-12], or above.

Reviewer: ```suggestion
A Schema MUST be represented by a JSON Schema document according to draft version 2020-12 [@JSON.SCHEMA.2020-12] or above.
```

> The Schema of a Verifiable Credential MUST include all properties that are required by this specification and MUST NOT change their cardinality.

Reviewer 2: must not change their type either or anything else that might affect their syntactic validity or semantic intent

Author: Can a JSON schema change the type of a credential? on the other hand, it is a good point that technically, it is possible to assert a specific `vct` value in the JSON schema which would be weird as it would prevent other types to extend that type, right?

Reviewer 2: I meant the data type where the "their" seemed to refer to any particular property. But what you said too.

```
{
  "$schema":"https://json-schema.org/draft/2020-12/schema",
  "type":"object",
  "properties":{
    "vct":{
      "type":"string"
    },
    "iss":{
      "type":"string"
    },
    "nbf":{
      "type":"number"
    },
    "exp":{
      "type":"number"
    },
    "cnf":{
      "type":"object"
    },
    "status":{
      "type":"object"
    },
    "given_name":{
      "type":"string"
    },
    "family_name":{
      "type":"string"
    },
    "email":{
      "type":"string"
    },
    "phone_number":{
      "type":"string"
    },
    "address":{
      "type":"object",
      "properties":{
        "street_address":{
          "type":"string"
        },
        "locality":{
          "type":"string"
        },
        "region":{
          "type":"string"
        },
        "country":{
          "type":"string"
        }
      }
    },
    "birthdate":{
      "type":"string"
    },
    "is_over_18":{
      "type":"boolean"
    },
    "is_over_21":{
      "type":"boolean"
    },
    "is_over_65":{
      "type":"boolean"
    }
  },
  "required":[
    "iss",
    "vct",
    "cnf"
  ]
}
```

> Note, that `iss` and `vct` are always required by this specification.

Reviewer: ```suggestion
Note that `iss` and `vct` are always required by this specification.
```

### Schema Validation {#schema-validation}

If a `schema` or `schema_uri` property is present, a Consumer MUST validate the JSON document resulting from the SD-JWT verification algorithm

> (as outlined in Section 8 of [@!I-D.ietf-oauth-selective-disclosure-jwt]) against the JSON Schema document provided by the `schema` or `schema_uri` property.

Reviewer: ```suggestion
(as defined in Section 8 of [@!I-D.ietf-oauth-selective-disclosure-jwt]) against the JSON Schema document provided by the `schema` or `schema_uri` property.
```
